controllers/topology/controller.go

```diff
@@ -117,32 +118,39 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_
 		return ctrl.Result{}, nil
 	}
 
+	// Create a scope initialized with only the cluster; during reconcile
+	// additional information will be added about the Cluster blueprint, current state and desired state.
+	scope := scope.NewScope(cluster)
+
 	// Handle normal reconciliation loop.
-	return r.reconcile(ctx, cluster)
+	return r.reconcile(ctx, scope)
 }
 
 // reconcile handles cluster reconciliation.
-func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
-	// Gets the ClusterClass and the referenced templates.
-	class, err := r.getClass(ctx, cluster)
+func (r *ClusterReconciler) reconcile(ctx context.Context, scope *scope.Scope) (ctrl.Result, error) {
+	var err error
+
+	// Gets the blueprint with the ClusterClass and the referenced templates
+	// and store it in the request scope.
+	scope.Blueprint, err = r.getBlueprint(ctx, scope)
 	if err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "error reading the ClusterClass")
 	}
 
-	// Gets the current state of the Cluster.
-	currentState, err := r.getCurrentState(ctx, cluster, class.clusterClass)
+	// Gets the current state of the Cluster and store it in the request scope.
+	scope.Current, err = r.getCurrentState(ctx, scope)
 	if err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "error reading current state of the Cluster topology")
 	}
 
-	// Computes the desired state of the Cluster
-	desiredState, err := r.computeDesiredState(ctx, class, currentState)
+	// Computes the desired state of the Cluster and store it in the request scope.
+	scope.Desired, err = r.computeDesiredState(ctx, scope)
 	if err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "error computing the desired state of the Cluster topology")
 	}
```

Comment:
Could we build these in the scope package instead?

Reply:
This PR is already huge.
Could we reconsider this in a separated PR (I'm not fully convinced moving all the code in the scope provides concrete benefits in code organisation.
Also, there is work to be done to add to the scope all the necessary machinery for avoiding orphaned objects...

Reply:
From a purely code perspective, this code is defining types in a package with a `New()` function that doesn't do much, and then we change the fields on the caller. In general, this isn't well contained behavior and would want to consider moving at least how we compose the scope to its respective package.

We can do this in another PR